umbral/pre.py

```diff
@@ -516,3 +486,27 @@ def decrypt(ciphertext: bytes, capsule: Capsule, decrypting_key: UmbralPrivateKe
         raise UmbralDecryptionError() from e
 
     return cleartext
+
+
+def decrypt_reencrypted(ciphertext: bytes,
+                        capsule: PreparedCapsule,
+                        cfrags: Sequence[CapsuleFrag],
+                        decrypting_key: UmbralPrivateKey,
+                        check_proof: bool = True) -> bytes:
+
+    if not isinstance(ciphertext, bytes) or len(ciphertext) < DEM_NONCE_SIZE:
+        raise ValueError("Input ciphertext must be a bytes object of length >= {}".format(DEM_NONCE_SIZE))
+    elif not isinstance(capsule, PreparedCapsule) or not capsule.verify():
+        raise Capsule.NotValid
+    elif not isinstance(decrypting_key, UmbralPrivateKey):
+        raise TypeError("The decrypting key is not an UmbralPrivateKey")
```

Comment:
These validation steps seems common to both decryption methods.

Reply:
I do feel a mild annoyance repeating this code, but it's quite trivial and short, so when it comes to choosing between repeating it and doing dynamic dispatch in `decrypt()` depending on whether we're doing it at Alice's or Bob's side, I choose the former.

In fact, I am not convinced that `isinstance()` checks are even necessary here. Feels somehow un-Pythonic. Could we just let it fail in `_decapsulate`?

Reply:
> when it comes to choosing between repeating it and doing dynamic dispatch in decrypt() depending on whether we're doing it at Alice's or Bob's side, I choose the former.

The fact that you have to choose is an indication this logic wants to be shared.  Why are we yanking it apart?

> Could we just let it fail in _decapsulate?

There we agree.

Reply:
> The fact that you have to choose is an indication this logic wants to be shared. Why are we yanking it apart?

One part of the shared logic is the `isinstance` checks, which, as you agree, we can get rid of. The other part (creating a DEM and using it) can be extracted into a function. I think that would solve the DRY problem.

As for why I'm yanking it apart, we essentially have two functions:
```
    decrypt(ciphertext, capsule, decrypting_key)
```
and
```
    decrypt(ciphertext, capsule, verifying_metadata, cfrags, decrypting_key, check_proof)
```
That seems like too much of a difference to me to justify putting them in a single method and checking if the user provided the right set of arguments.

Reply:
Having said all that, this change is not critical to the purpose of the PR, and if it causes controversy, I can get rid of it.

Reply:
Got rid of code duplication by removing type checks and moving verification checks to where they belong.